### PR TITLE
[SPARK-36906][PYTHON] Inline type hints for conf.py and observation.py in python/pyspark/sql

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Inline type hints for conf.py and observation.py in python/pyspark/sql.

### Why are the changes needed?
Currently, there is type hint stub files (*.pyi) to show the expected types for functions, but we can also take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No.

It has a DOC typo fix:
`Metrics are aggregation expressions, which are applied to the DataFrame while **is** is being`
is changed to
`Metrics are aggregation expressions, which are applied to the DataFrame while **it** is being`

### How was this patch tested?
Existing test.

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

File: python/pyspark/sql/conf.py [hunk 1]
```diff
@@ -26,17 +29,17 @@ class RuntimeConfig(object):
...
     """
 
-    def __init__(self, jconf):
+    def __init__(self, jconf: JavaObject) -> None:
         """Create a new RuntimeConfig that wraps the underlying JVM object."""
         self._jconf = jconf
 
     @since(2.0)
-    def set(self, key, value):
+    def set(self, key: str, value: str) -> None:
```
Comment (zero323) [inline python/pyspark/sql/conf.py:37]: Seems like this can be more permissive on `value` type (`Union[str, int, bool]`?)

**Review by zero323 — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by ueshin — APPROVED** [review]
LGTM.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**